Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs

```diff
@@ -1407,5 +1412,126 @@ public void OnSpeechKeywordRecognized(SpeechEventData eventData)
             gazePointerStateMachine.OnSpeechKeywordRecognized(eventData);
         }
         #endregion
+
+
+        #region IPointerPreferences Implementation
+        private List<PointerPreferences> customPointerBehaviors = new List<PointerPreferences>();
+
+        /// <inheritdoc />
+        public PointerBehavior GetPointerBehavior(IMixedRealityPointer pointer)
+        {
+            // Assumption: all pointers have controllers, input sources, except the gaze pointers
+            // if the controller, input source is null, return the gaze pointer behavior here.
+            if (pointer.Controller == null || pointer.InputSourceParent == null)
+            {
+                // gazepointer means input source is null
+                return GazePointerBehavior;
+            }
+
+            return GetPointerBehavior(
+                pointer.GetType(), 
+                pointer.Controller.ControllerHandedness,
+                pointer.InputSourceParent.SourceType);
+        }
+
+        /// <summary>
+        /// Gets the behavior for the given pointer type.
+        /// </summary>
+        /// <param name="pointerType">Pointer type to query</param>
+        /// <param name="handedness">Handedness to query</param>
+        /// <returns><seealso cref="Microsoft.MixedReality.Toolkit.Input.PointerBehavior"/> for the given pointer type and handedness. If right hand is enabled, left
+        /// hand is not enabled, and Handedness.Any is passed, returns value for the right hand.</returns>
+        public PointerBehavior GetPointerBehavior<T>(
+            Handedness handedness,
+            InputSourceType sourceType) where T : class, IMixedRealityPointer
+        {
+            return GetPointerBehavior(typeof(T), handedness, sourceType);
+        }
+
+        private PointerBehavior GetPointerBehavior(Type type, Handedness handedness, InputSourceType sourceType)
```

Comment:
>GetPointerBehavior [](start = 32, length = 18)

If we expect that GetPointerBehaviors can be called in the Update() loop, consider avoiding LINQ:

[https://docs.microsoft.com/en-us/windows/mixed-reality/performance-recommendations-for-unity#avoid-expensive-operations](https://docs.microsoft.com/en-us/windows/mixed-reality/performance-recommendations-for-unity#avoid-expensive-operations)

Reply:
Same for SetPointerBehavior

---
In reply to: [323507446](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5920#discussion_r323507446) [](ancestors = 323507446)